### PR TITLE
Add neural networks and bug fixes

### PR DESCRIPTION
**Pull request by MikhailKuz**
## Additions
* Neural networks in tabular preset (feature pipelines, models, tuning utils, grad clipping for regularization)
* Tutorial for nn and link for Kaggle competition
* Loss on logits for linear model (fixes sb-ai-lab/LightAutoML/issues/75)

## Fixes
* [kfold and reader seeds](https://github.com/sb-ai-lab/LightAutoML/pull/79/commits/501f4c08fc94ac9821d8fb81cc9806b21efe8143) - if the sample (100k) is less than the number of observations per fold, then it samples the observations from only one fold (since folds are sorted). So the multiclass (and multioutput) target encoder with division by zero falls.
* [deprecated series append](https://github.com/sb-ai-lab/LightAutoML/pull/79/commits/e6dd5c0b6ff52b7fe33cd021845d700d8a3bd135) - verbose deprecation warning (fixes sb-ai-lab/LightAutoML/issues/56)
* [verbose](https://github.com/sb-ai-lab/LightAutoML/pull/79/commits/bcb56dba43c8d40ba4ed8c4f430d7589e502e5a0) - get right verbose level from verbose in `automl.fit_predict`
* [weight in metric func](https://github.com/sb-ai-lab/LightAutoML/pull/79/commits/b197395e8125eec8c46f3c4bf26fd0c726014b16) - make available use of custom weight instead of default 1 in neural networks
* [fix cb custom multitask metric](https://github.com/sb-ai-lab/LightAutoML/pull/79/commits/d2c066e4569f1f4e1203cdcc252ea90872d42e29) - passed one-dim version of losses instead of multi-dim version (fixes sb-ai-lab/LightAutoML/issues/74)

### CODE REVIEW

File: lightautoml/ml_algo/torch_based/nn_models.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+        ],
+        drop_rate: Union[float, List[float]] = 0.1,
+        act_fun: nn.Module = nn.ReLU,
+        noise_std: float = 0.05,
+        num_init_features: Optional[int] = None,
+        use_bn: bool = True,
+        use_noise: bool = False,
+        concat_input: bool = True,
+        device: torch.device = torch.device("cuda:0"),
+        **kwargs,
```
Comment (dev-rinchin) [inline lightautoml/ml_algo/torch_based/nn_models.py:120]: Do you need kwargs here?

File: lightautoml/ml_algo/torch_based/nn_models.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        Args:
+            stddev: Std of noise.
+            device: Device to compute on.
+
+        """
+        super().__init__()
+        self.stddev = stddev
+        self.device = device
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (dev-rinchin) [inline lightautoml/ml_algo/torch_based/nn_models.py:47]: Add docstring in all public methods

File: lightautoml/ml_algo/torch_based/nn_models.py (hunk 1 above)
Comment (MikhailKuz) [reply lightautoml/ml_algo/torch_based/nn_models.py:120]: If you want to use a custom model, then params are passed to the creation method. So if we remove kwargs here, there will be no availability to use custom model (tuning your params for example).

**Review by dev-rinchin — COMMENTED** [review]
LGTM

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 3]
```diff
@@ -0,0 +1,1146 @@
...
+    "print(automl.create_model_str_desc())"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "8a26dc61",
+   "metadata": {},
+   "source": [
+    "## 4. Available built-in models"
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:458]: I'd like to place this section before section with TabularAutoML init

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 4]
```diff
@@ -0,0 +1,1146 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "b8c3218d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "DATASET_DIR = '../data/'\n",
+    "DATASET_NAME = 'sampled_app_train.csv'\n",
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:230]: remove redundant variables DATASET_NAME and DATASET_DIR -- they are used just only once

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 5]
```diff
@@ -0,0 +1,1146 @@
...
+     "exception": false,
+     "start_time": "2021-06-22T20:10:53.024294",
+     "status": "completed"
+    },
+    "tags": []
+   },
+   "source": [
+    "### 0.1. Import libraries\n",
+    "\n",
+    "Here we will import the libraries we use in this kernel:\n",
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:104]: Redundant comments in lines 105-107

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 6]
```diff
@@ -0,0 +1,1146 @@
...
+    "To use different model pass it to the list in `\"use_algo\"`. We support custom models inherited from `torch.nn.Module` class. For every model their parameters is listed below."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "0ecdf10d",
+   "metadata": {},
+   "source": [
+    "### 4.1. MLP (`\"mlp\"`)\n",
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:476]: please, add links to description of NN architecture for every section below

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 7]
```diff
@@ -0,0 +1,1146 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "8c7cd871",
+   "metadata": {},
+   "source": [
+    "- `bs` - batch_size\n",
+    "- `snap_params` - early stopping and checkpoint averaging params, swa\n",
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:535]: add definition for "swa"

File: examples/tutorials/Tutorial_9_neural_networks.ipynb [hunk 8]
```diff
@@ -0,0 +1,1146 @@
...
+    ")\n",
+    "automl.fit_predict(tr_data, roles=roles, verbose=1)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bbf8a589",
+   "metadata": {},
+   "source": [
+    "#### 6.1.2 Define by yourself whole pipeline"
```
Comment (dev-rinchin) [inline examples/tutorials/Tutorial_9_neural_networks.ipynb:726]:
```suggestion
    "#### 6.1.2 Define the pipeline by yourself"
```

File: pyproject.toml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "LightAutoML"
-version = "0.3.7.3"
+version = "0.3.7.4"
```
Comment (dev-rinchin) [inline pyproject.toml:3]:
```suggestion
version = "0.3.7.3"
```

Comment (dev-rinchin) [reply pyproject.toml:3]: version should be updated only once exactly before release

File: lightautoml/automl/base.py [hunk 10]
```diff
@@ -118,6 +120,7 @@ def _initialize(
                 input features to next levels.
             return_all_predictions: True if we should return all predictions from last
                 level models.
+            debug: To catch running model exceptions or not.
```
Comment (dev-rinchin) [inline lightautoml/automl/base.py:123]: add more information about behaviour, for example:
True: show exceptions during model training;
False: catch and hide exceptions;

File: lightautoml/automl/presets/tabular_presets.py [hunk 11]
```diff
@@ -453,6 +538,16 @@ def create_automl(self, **fit_args):
...
+            available_nn_models = available_nn_models + [x + "_tuned" for x in available_nn_models]
+            nn_models = [
+                x for x in names if x in available_nn_models or (not isinstance(x, str) and issubclass(x, nn.Module))
+            ]
+
+            if len(nn_models) > 0:
+                selector = None
+                lvl.append(self.get_nn(nn_models, n + 1, selector))
+
             levels.append(lvl)
```
Comment (dev-rinchin) [inline lightautoml/automl/presets/tabular_presets.py:551]: do not append lvl if it's empty

File: lightautoml/automl/presets/tabular_presets.py [hunk 12]
```diff
@@ -453,6 +538,16 @@ def create_automl(self, **fit_args):
                     selector = pre_selector
                 lvl.append(self.get_gbms(gbm_models, n + 1, selector))
 
+            available_nn_models = ["nn", "mlp", "dense", "denselight", "resnet", "snn", "linear_layer", "_linear_layer"]
+            available_nn_models = available_nn_models + [x + "_tuned" for x in available_nn_models]
+            nn_models = [
+                x for x in names if x in available_nn_models or (not isinstance(x, str) and issubclass(x, nn.Module))
```
Comment (dev-rinchin) [inline lightautoml/automl/presets/tabular_presets.py:544]: x can't be a issubclass of nn.Module and an instance or str at the same time

File: lightautoml/text/nn_model.py [hunk 13]
```diff
@@ -291,27 +301,42 @@ def __init__(
...
-        x = self.predict(inp)
-        loss = self.loss(inp["label"].view(inp["label"].shape[0], -1), x, inp.get("weight", None))
-        return loss
-
-    def predict(self, inp: Dict[str, torch.Tensor]) -> torch.Tensor:
-        """Prediction."""
+            try:
+                last_layer = list(
+                    filter(
+                        lambda x: isinstance(x, nn.Linear) or isinstance(x, nn.Sequential),
```
Comment (DESimakov) [inline lightautoml/text/nn_model.py:319]: Remove all lambda functions.

Check whether you can load automl with nn on disk and load it.
